### DISCUSSION TITLE
How should the modular and non modular synthax live with current react-native-firebase version ?

### DISCUSSION THREAD

**Discussion by kopax-polyconseil** [invertase/react-native-firebase#6220] [Q&A] [answered]
Hello Guys, 

I am upgrading my firebase web version in a `react-native-web` application and I am now wondering how the two synthax (modalar and not modular) should live together ? 

Do you have any plans to upgrade the synthaxe for react native firebase ?

For instance this is an example of new synthax for firestore:

```jsx
const firestore = getFirestore(app);
const docRef = doc(firestore, 'doc/doc');
const docSnap = await getDoc(docRef);
console.log('docSnap', {
  docSnap
})
await updateDoc(docRef, "lucas", 'lucasbis');
console.log('ending firebase store')
```

This is the old synthax (non modular) used by react-native-firebase 

```jsx
firestoreRemoteStore
.collection(RemoteStoreCollections.APPLICATION_VERSIONS)
.doc(env.ENV)
.onSnapshot(
  (docSnapshot) => {
    onValueChange(docSnapshot.get(RemoteStoreDocuments.MINIMAL_BUILD_NUMBER))
  },
  (error) => {
    captureMonitoringError(error.message, 'firestore_not_available')
  }
)
```

Thanks in advance for clarifying !

**Comment by mikehardy** [comment, marked as answer]
The two syntaxes are simply not compatible and you may not mix them.
Until we support the v9 web syntax here (it is on the roadmap, but is not done yet...) you will have to use the firebsae-js-sdk either v8 or compat interfaces

**Reply by mikehardy** [reply]
You can see how I shim the firebase-js-sdk into a react-native-web project using the compat APIs here https://github.com/invertase/react-native-firebase-authentication-example/tree/main/template/src/shims

**Reply by kopax-polyconseil** [reply]
Thanks a lot for the quick reply, do you have any ETA on when modular mode will be available ? 

I'll accept this as an answer but thanks in advance for any additional claryfication that can help up schedule the upgrade later.

And thanks a lot for sharing this lib, have a great day !

**Reply by mikehardy** [reply]
I do not have an ETA, I'm sorry. At this point it is looking like it may be 2022Q3, but that's earliest. There are some uninteresting but really important things we need to handle with regards to our react-native testing infrastructure, plus some necessary forward-ports to an upcoming firebase-ios-sdk breaking change before we can do other stuff

**Reply by Nantris** [reply]
>  At this point it is looking like it may be 2022Q3, but that's earliest.

Thank you for all your hard work; very much looking forward to this! Do you have any updated idea of a timeline?

**Reply by mikehardy** [reply]
No update sorry - "that's earliest" was factual, and I still don't have an open slot right now for implementation. I wish it were either done already or I had open schedule and was already deep in to it, but wishes aren't code alas

**Reply by Nantris** [reply]
I know all too well what that's like. I'm super appreciative for all you do to make this project possible, so thanks even for the reply!

I figured it was 2023 at earliest when I asked, but I mean like, 2023 first half, second half, or are we looking at even beyond that? I'm just trying to figure out a guesstimate for an internal roadmap on our end.

**Reply by mikehardy** [reply]
I just can't predict - I wish I could, if you are bridging react-native-firebase and firebase-js-sdk as I do occasionally: 

https://github.com/invertase/react-native-firebase-authentication-example/tree/main/template/src/shims
https://github.com/invertase/react-native-firebase-authentication-example/blob/68bd37f8c219f892655b40c73e9818370f368b82/template/craco.config.js#L111-L114

I think you may use web v9 - https://github.com/invertase/react-native-firebase-authentication-example/blob/68bd37f8c219f892655b40c73e9818370f368b82/template/package.json#L40

You just have to do the compat interfaces: https://github.com/invertase/react-native-firebase-authentication-example/blob/68bd37f8c219f892655b40c73e9818370f368b82/template/src/shims/firebase-auth-web.ts#L1-L2

As those are actually working for me there is less pressure to do this repo-wide change when I'm already strapped for time.
At the same time, obviously modular is the future, I'm not arguing that in any way

Currently sometime in 2023 sounds about right :sweat_smile:

**Reply by Nantris** [reply]
> As those are actually working for me there is less pressure to do this repo-wide change when I'm already strapped for time.

That's totally fair! We're a definite edge case where we use shared Firebase code across mobile and web app, and the web app requires the modular syntax so the shims won't help us, but that's awesome that they work for most use cases! There are (or at least were) ongoing issues with the compatibility layer too - which is why we had to ditch it.

But hey, we can all still use `firebase-js-sdk` in the meanwhile if needed so yeah, as much as I'd love to have this wishlist item sooner, it's not urgent.

Thanks very much for taking the time to write that all up!

**Reply by mikehardy** [reply]
Oh I'm sorry to hear the compat issues specifically impacted your project. I had heard there were some but of course "works for me" and all. That's a good data point to know though. It does need to happen, hopefully before we all go around the sun again

**Reply by Ehesp** [reply]
Probably worth a mention that there is a number of considerations we need to think about too:

- We'd probably need to roll both version alongside each other for a period of time, otherwise it'd be a total "you need to fully overhaul your project to keep using newer versions of RNFB". This just adds a huge maintenance overhead.
- Documentation would need a full overhaul, given we have examples and reference API. We also currently generate docs from the reference API, so no idea how that would work with the modular approach. The docs site is a little bit brittle / specifically created to handle this.
- With a dual approach, we'd need to (probably) create new npm packages OR major version the modular ones under beta whilst also pushing out updates to the older packages - this sounds a bit messy.

I'm also not sure we could do a 1-1 mapping of the JS SDK and RNFB, given there are APIs which either don't exist on web, or would have to be different (e.g. firestore server timestamp behaviour). Not entirely sure how this would fit together right now.

**Reply by mikehardy** [reply]
All totally valid of course. But a vN+1@beta then just cut it wouldn't be that hard. I'm not saying that to be dismissive, I just converted all my functions to the new API (firebase-admin uses modular style) and it was trivial. 

Could probably write a codemod. 

And our release process supports releasing on branches now (implemented + tested with the v14 series to solve a compatibility issue with react-native 0.70 for folks still struggling with use_frameworks)

So I think it's more "emergency fixes on branch for non-modular / just cut a major version + write a codemod for modular and move on"

With regards to docs, I can't speak to that. Current docs site cannot be upgraded for a variety of dependency-related issues so perhaps better to get docs.page working with typedoc and just go with that :)

**Reply by Nantris** [reply]
> We'd probably need to roll both version alongside each other for a period of time, otherwise it'd be a total "you need to fully overhaul your project to keep using newer versions of RNFB". This just adds a huge maintenance overhead.

Firebase itself adds a compatibility layer instead between JS SDK v8 and v9. I don't know if that's plausible though. 

I'd be impressed if a codemod could really do the heavy lifting for the upgrade.

**Reply by Ehesp** [reply]
The compatibility layer won't be kept up-to-date with v9 from what I understand, it'll eventually stop receiving updates.

We'll have some proper discussions internally and sync with the Firebase team to better decide here. I think the general gist is that it's totally possible, just it comes with a huge amount of work. v9 API was created for tree shaking, but AFAIK that's not really a problem here due to the thin JS abstraction layer of RNFB. The real topic is whether to keep in sync with the v9 public API as closely as we can.

**Reply by Nantris** [reply]
> The compatibility layer won't be kept up-to-date with v9 from what I understand, it'll eventually stop receiving updates.

That seems totally reasonable both in terms of Firebase and RNF

FWIW, I think the API should match the modular API as that's the future of Firebase. But perhaps I'm biased, because we won't be able to use this library in the future if the API differs from the v9 API, unless it's a tiny difference here or there. Our code is cross-platform and we were hoping to use React-Native-Firebase as a drop-in replacement on React Native, while still allowing our Firebase v9 code to run elsewhere.

**Reply by Ehesp** [reply]
Fair points - I'll raise them up and we can get a proper discussion going.

QQ: in somecases a modular RNFB API would differ from that on Firebase v9 (not much, but there will be subtle changes due to the native layer). I suspect RNFB would have to proxy to the v9 SDK internally, rather than aliasing the RNFB import to the Firebase SDK directly?

**Reply by mikehardy** [reply]
There are already subtle difference, but the ones I see are usually around returns, that is perhaps RNFB returns Promise whereas firebase-js-sdk may not. Those are non-events as await-ing a non-Promise doesn't matter. I haven't seen other differences that matter much, but in the event that there is some real difference I think job 1 will be to try to remove the difference ;-), but okay if not possible then there will need to be a proxy of some sort.

Perhaps we can ship a set of shims that import firebase-js-sdk and for those differences we implement small compatibility changes prior to re-exporting. That's how I handle it in my react-native-web + RNFB app

For instance, in-app-messaging simply does not exist on the firebase-js-sdk (yet?) but my app uses it specifically to suppress messages at times, so I have a shim that exports dummy APIs like so:

```typescript
const inAppMessaging = () => {
  return {
    setMessagesDisplaySuppressed: async (_: boolean): Promise<void> => {},
  };
};
export default inAppMessaging;
```

Same with crashlytics, I have in my react-native package `./src/shims/firebase-crashlytics-web.ts` and it is loaded into webpack:

```typescript
const crashlytics = () => {
  return {
    setUserId: (_: string): void => {},
    log: (_: string): void => {},
  };
};
export default crashlytics;
```
